File: day5/sol.py
from joblib import Parallel, delayed


def part1():
    with open("data.txt") as file:
        lines = file.readlines()

    seeds = list(map(int, lines[0][lines[0].find(':')+1:].split()))

    tables = []
    l = 3
    while l < len(lines):
        tables.append([])
        while l < len(lines) and lines[l][0] != '\n':
            tables[-1].append(tuple(map(int, lines[l].split())))
            l += 1
        l += 2

    min_loc = -1

    for seed in seeds:
        x = seed
        for table in tables:
            for entry in table:
                if entry[1] <= x < entry[1] + entry[2]:
                    x = entry[0] + x - entry[1]
                    break

        if min_loc < 0 or x < min_loc:
            min_loc = x

    print(f"Best location = {min_loc}")


def process(i, seeds, tables):
    min_loc = -1
    beg, end = seeds[i]
    per = 0
    for seed in range(beg, beg + end):
        loc = seed
        for table in tables:
            for entry in table:
                if entry[1] <= loc < entry[1] + entry[2]:
                    loc = entry[0] + loc - entry[1]
                    break
        if min_loc < 0 or loc < min_loc:
            min_loc = loc
        if (seed + 1 - beg) * 100 // end > per:
            per += 1
            for _ in range(i):
                print(end="     ")
            print(f"{per:3.0f}% ", end="")
            for _ in range(i + 1, len(seeds)):
                print(end="     ")
            print()

    return min_loc


def part2():
    with open("data.txt") as file:
        lines = file.readlines()

    seeds = list(map(int, lines[0][lines[0].find(':')+1:].split()))
    seeds = [(seeds[2 * i], seeds[2 * i + 1]) for i in range(len(seeds) // 2)]

    tables = []
    l = 3
    while l < len(lines):
        tables.append([])
        while l < len(lines) and lines[l][0] != '\n':
            tables[-1].append(tuple(map(int, lines[l].split())))
            l += 1
        l += 2

    min_locs = Parallel(n_jobs=len(seeds))(
        delayed(process)(i, seeds, tables) for i in range(len(seeds)))
    min_loc = min(min_locs)
    print()

    print(f"Best location = {min_loc}")


def main():
    part1()
    part2()


if __name__ == "__main__":
    main()

File: day5/sol.cpp
#include <algorithm>
#include <array>
#include <charconv>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <omp.h>
#include <vector>

int main()
{
    uint64_t min_loc = ~0U;

    std::ifstream file{"data.txt"};
    file.seekg(0, std::ios::end);
    std::string data(file.tellg(), 0);
    file.seekg(0, std::ios::beg);
    file.read(data.data(), data.size());

    std::vector<uint64_t> seeds;
    size_t c = 0;

    while (data[c] != ':')
        ++c;
    ++c;

    while (data[c] != '\n')
    {
        size_t i = ++c;
        while (std::isdigit(data[c]))
            ++c;

        std::from_chars(&data[i], &data[c], seeds.emplace_back());
    }
    seeds.shrink_to_fit();

    std::array<std::vector<std::array<uint64_t, 3>>, 7> tables;

    for (size_t i = 0; i < tables.size(); ++i)
    {
        size_t j;

        c += 2;
        while (data[c] != '\n')
            ++c;

        do
        {
            tables[i].emplace_back();

            j = ++c;
            while (std::isdigit(data[c]))
                ++c;
            std::from_chars(&data[j], &data[c], tables[i].back()[0]);

            j = ++c;
            while (std::isdigit(data[c]))
                ++c;
            std::from_chars(&data[j], &data[c], tables[i].back()[1]);

            j = ++c;
            while (std::isdigit(data[c]))
                ++c;
            std::from_chars(&data[j], &data[c], tables[i].back()[2]);
        } while (data[c + 1] != '\n');
    }

    for (auto &&seed : seeds)
    {
        uint64_t loc = seed;

        for (auto &&table : tables)
            for (auto &&[d, s, r] : table)
                if (s <= loc && loc < s + r)
                {
                    loc = d + loc - s;
                    break;
                }

        if (loc < min_loc)
            min_loc = loc;
    }

    std::cout << "Best location = " << min_loc << '\n';

    int num_threads = seeds.size() >> 1;
    omp_set_num_threads(num_threads);
    std::vector<uint64_t> min_locs(num_threads, ~0U);

#pragma omp parallel
    {
        int i = omp_get_thread_num();
        uint64_t start = seeds[2 * i];
        uint64_t n = seeds[2 * i + 1];
        uint64_t per = 0;
        std::string prefix;
        std::string suffix;

        for (size_t j = 0; j < i; ++j)
            prefix += "     ";

        for (size_t j = i + 1; j < num_threads; ++j)
            suffix += "     ";

        for (uint64_t seed = start; seed < start + n; ++seed)
        {
            uint64_t loc = seed;

            for (auto &&table : tables)
                for (auto &&[d, s, r] : table)
                    if (s <= loc && loc < s + r)
                    {
                        loc = d + loc - s;
                        break;
                    }

            if (loc < min_locs[i])
                min_locs[i] = loc;
            if ((seed + 1 - start) * 100 / n > per)
            {
                ++per;
#pragma omp critical
                {
                    std::cout << prefix << std::setw(3) << per << "% " << suffix << '\n';
                    std::cout.flush();
                }
            }
        }
    }
    std::cout << '\n';

    min_loc = *std::ranges::min_element(min_locs);

    std::cout << "Best location = " << min_loc << '\n';

    return 0;
}
